x-pack/plugins/ml/public/application/components/model_snapshots/revert_model_snapshot_flyout/create_calendar.tsx

```diff
@@ -0,0 +1,289 @@
+                <EuiFlexGroup>
+                  <EuiFlexItem>
+                    <EuiFormRow
+                      fullWidth
+                      label={i18n.translate(
+                        'xpack.ml.revertModelSnapshotFlyout.createCalendar.descriptionLabel',
+                        {
+                          defaultMessage: 'Description',
+                        }
+                      )}
+                    >
+                      <EuiFieldText
+                        style={{ width: '100%', maxWidth: 'inherit' }}
+                        value={c.description}
+                        onChange={(e) => setDescription(e.target.value, i)}
+                      />
+                    </EuiFormRow>
+                  </EuiFlexItem>
+                </EuiFlexGroup>
+              </EuiFlexItem>
+              <EuiFlexItem
+                grow={false}
+                style={{
+                  borderLeft: `1px solid ${euiTheme.euiColorLightShade}`,
+                  marginRight: '0px',
+                }}
+              />
+              <EuiFlexItem grow={false}>
+                <EuiButtonIcon
+                  style={{ margin: 'auto' }}
+                  color={'danger'}
+                  onClick={() => removeCalendarEvent(i)}
+                  iconType="trash"
+                  aria-label={i18n.translate(
+                    'xpack.ml.revertModelSnapshotFlyout.createCalendar.deleteLabel',
+                    {
+                      defaultMessage: 'Delete event',
+                    }
+                  )}
+                />
+              </EuiFlexItem>
+            </EuiFlexGroup>
+          </EuiPanel>
+          <EuiSpacer size="m" />
+        </div>
+      ))}
+    </>
+  );
+};
+
+interface ChartProps {
+  eventRateData: LineChartPoint[];
+  anomalies: Anomaly[];
+  loading: boolean;
+  onBrushEnd(area: XYBrushArea): void;
+  overlayRanges: Array<{ start: number; end: number }>;
+}
+
+const Chart: FC<ChartProps> = memo(
+  ({ eventRateData, anomalies, loading, onBrushEnd, overlayRanges }) => (
+    <EventRateChart
+      eventRateChartData={eventRateData}
+      anomalyData={anomalies}
+      loading={loading}
+      height={'100px'}
+      width={'100%'}
+      fadeChart={true}
+      overlayRanges={overlayRanges.map((c) => ({
+        start: c.start,
+        end: c.end,
+        color: '#0000ff',
+        showMarker: false,
+      }))}
+      onBrushEnd={onBrushEnd}
+    />
+  ),
+  (prev: ChartProps, next: ChartProps) => {
+    // only redraw if the calendar ranges have changes
+    return (
+      prev.overlayRanges.length === next.overlayRanges.length &&
+      JSON.stringify(prev.overlayRanges) === JSON.stringify(next.overlayRanges)
+    );
+  }
+);
+
+function filterIncompleteEvents(event: CalendarEvent): event is CalendarEvent {
+  return event.start !== null && event.end !== null;
+}
+
+function createDefaultEventDescription(index: number) {
+  return `Auto created event ${index}`;
+}
```

Comment:
These functions are only used inside of the component, hence they should be defined inside

Reply:
as they don't rely on state, i moved them out of the component to avoid redefining on render.